lib/utils/promote.ts

```diff
@@ -0,0 +1,371 @@
+	await assertDeviceIsCompatible(deviceIp);
+	logger.logDebug(`Using device: ${deviceIp}`);
+
+	logger.logDebug('Deconfiguring...');
+	await deconfigure(deviceIp);
+	logger.logDebug('All done.');
+
+	const platformUrl = await sdk.settings.get('resinUrl');
+	logger.logSuccess(`Device successfully left ${platformUrl}.`);
+}
+
+export class ExecError extends TypedError {
+	public cmd: string;
+	public exitCode: number;
+
+	constructor(cmd: string, exitCode: number) {
+		super(`Command '${cmd}' failed with error: ${exitCode}`);
+		this.cmd = cmd;
+		this.exitCode = exitCode;
+	}
+}
+
+async function exec(
+	deviceIp: string,
+	cmd: string,
+	stdout?: NodeJS.WritableStream,
+): Promise<void> {
+	const command = `ssh \
+		-t \
+		-p 22222 \
+		-o LogLevel=ERROR \
+		-o StrictHostKeyChecking=no \
+		-o UserKnownHostsFile=/dev/null \
+		root@${deviceIp} \
+		${cmd}`;
+
+	const stdio = ['ignore', stdout ? 'pipe' : 'inherit', 'ignore'];
+	const { program, args } = getSubShellCommand(command);
+
+	const exitCode = await new Bluebird<number>((resolve, reject) => {
+		const ps = spawn(program, args, { stdio })
+			.on('error', reject)
+			.on('close', resolve);
+
+		if (stdout) {
+			ps.stdout.pipe(stdout);
+		}
+	});
+	if (exitCode != 0) {
+		throw new ExecError(cmd, exitCode);
+	}
+}
+
+async function execBuffered(
+	deviceIp: string,
+	cmd: string,
+	enc?: string,
+): Promise<string> {
+	const through = await import('through2');
+	const buffer: string[] = [];
+	await exec(
+		deviceIp,
+		cmd,
+		through(function(data, _enc, cb) {
+			buffer.push(data.toString(enc));
+			cb();
+		}),
+	);
+	return buffer.join('');
+}
+
+async function execCommand(
+	deviceIp: string,
+	cmd: string,
+	msg: string,
+): Promise<void> {
+	const through = await import('through2');
+	const visuals = await import('resin-cli-visuals');
+
+	const spinner = new visuals.Spinner(`[${deviceIp}] Connecting...`);
+	const innerSpinner = spinner.spinner;
+
+	const stream = through(function(data, _enc, cb) {
+		innerSpinner.setSpinnerTitle(`%s [${deviceIp}] ${msg}`);
+		cb(null, data);
+	});
+
+	spinner.start();
+	await exec(deviceIp, cmd, stream);
+	spinner.stop();
+}
+
+async function configure(deviceIp: string, config: any): Promise<void> {
+	const json = JSON.stringify(config);
+	const b64 = Buffer.from(json).toString('base64');
+	const str = `"$(base64 -d <<< ${b64})"`;
+	await execCommand(deviceIp, `os-config join '${str}'`, 'Configuring...');
+}
+
+async function deconfigure(deviceIp: string): Promise<void> {
+	await execCommand(deviceIp, 'os-config leave', 'Configuring...');
+}
+
+async function assertDeviceIsCompatible(deviceIp: string): Promise<void> {
+	const { exitWithExpectedError } = await import('../utils/patterns');
+	try {
+		await execBuffered(deviceIp, 'os-config --version');
+	} catch (err) {
+		exitWithExpectedError(stripIndent`
+			Device "${deviceIp}" is incompatible and cannot join or leave an application.
+			Please select or provision a newer device.`);
```

Comment:
Not yet unfortunately, the feature is still unreleased. I'll make sure to add version info before merging this (not much use of these commands without resinOS support anyway, so it may wait for a release).